### PR TITLE
2D flux field visualization

### PR DESCRIPTION
**Pull request by termi-official**
```julia
using FerriteViz
import GLMakie

include("docs/src/ferrite-examples/heat-equation.jl");
dh,u=manufactured_heat_problem(Triangle, Lagrange{2,RefTetrahedron,2}(), 21)
(dh_grad, u_grad) = FerriteViz.interpolate_gradient_field(dh, u, :u);

FerriteViz.surface(dh_grad, u_grad)
```

L2-gradient-field

Stepped up the game and checked analytically the gradient field. It converges. :)

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -424,3 +424,49 @@ function dof_to_node(dh::Ferrite.AbstractDofHandler, u::Array{T,1}; field::Int=1
...
+
+    for (cell_num, cell) in enumerate(Ferrite.CellIterator(dh))
+        Ferrite.celldofs!(cell_dofs, dh, cell_num)
+        uᵉ = u[cell_dofs[Ferrite.dof_range(dh, field_name)]]
+
+        Ferrite.celldofs!(cell_dofs_gradient, dh_gradient, cell_num)
+        uᵉ_gradient = u_gradient[cell_dofs_gradient]
+
+        ref_coords = Ferrite.reference_coordinates(ip_gradient)
+        for i ∈ 1:Ferrite.getnbasefunctions(ip_gradient)
```
Comment (koehlerson) [inline src/utils.jl:462]: Within this loop we need another for loop for the dimension of the field, something like this:
```
    for (cell_num, cell) in enumerate(Ferrite.CellIterator(dh))
        Ferrite.celldofs!(cell_dofs, dh, cell_num)
        uᵉ = u[cell_dofs[Ferrite.dof_range(dh, field_name)]]

        Ferrite.celldofs!(cell_dofs_gradient, dh_gradient, cell_num)
        uᵉ_gradient .= u_gradient[cell_dofs_gradient]

        ref_coords = Ferrite.reference_coordinates(ip_gradient)
        for i ∈ 1:Ferrite.getnbasefunctions(ip_gradient)
            _derivative = Ferrite.derivative(ip, ref_coords[i])
            for j ∈ 1:Ferrite.getnbasefunctions(ip)
                for d in 1:Ferrite.getfielddim(dh,field_name)
                    uᵉ_gradient[d*i] += uᵉ[j*d] * _derivative[j][d]
                end
            end 
        end
        u_gradient[cell_dofs_gradient] .= uᵉ_gradient
    end
    return (dh_gradient, u_gradient)
```
Either I messed something up with the indices access or maybe its by construction missing some dofs (probably the prior) ? The incompressible elasticity example with quadratic displacement looks like this:
image

The `getnbasefunctions` does not know about the dimensionality of the field and returns e.g. 

```julia
julia> Ferrite.getnbasefunctions(DiscontinuousLagrange{2,RefTetrahedron,1}())
3
```
which is true for scalar fields, for vector fields the dimensionality need to be taken into account e.g. `4` entries for 2D field in 2D space, which ends up in 12 dofs per cell

```julia
julia> Ferrite.getfielddim(dh_grad,:gradient)
4

julia> Ferrite.ndofs_per_cell(dh_grad)
12

```

File: src/utils.jl [hunk 2]
```diff
@@ -424,3 +424,49 @@ function dof_to_node(dh::Ferrite.AbstractDofHandler, u::Array{T,1}; field::Int=1
...
+    ip = Ferrite.getfieldinterpolation(dh, field_idx)
+
+    # Dof handler for gradient field
+    dh_gradient = Ferrite.DofHandler(getgrid(dh))
+    ip_gradient = get_gradient_interpolation(ip)
+    push!(dh_gradient, :gradient, Ferrite.getfielddim(dh, field_name)*dim, ip_gradient) # field dim × spatial dim components
+    Ferrite.close!(dh_gradient)
+
+    # Buffer for the dofs
+    cell_dofs = zeros(Int, Ferrite.getnbasefunctions(ip))
```
Comment (koehlerson) [inline src/utils.jl:448]:
```suggestion
    cell_dofs = zeros(Int, Ferrite.ndofs_per_cell(dh))
```

**Review by koehlerson — COMMENTED** [review]
nice, I think it would be nice to support vector valued fields in 2D for this.

Can you elaborate on https://github.com/Ferrite-FEM/FerriteViz.jl/blob/master/src/utils.jl#L292-L297 ?

**Comment by termi-official** [comment]
The problem mentioned in https://github.com/Ferrite-FEM/FerriteViz.jl/blob/master/src/utils.jl#L292-L297 is that for volumetric entities we need an interpolation on entities with dimension 2 (for the most common problems). Now, L2 spaces have support on the local portion of a face, but not on the actual shared face. However, the current implementation just extract the faces from the provided interpolation to build the face triangles - which is an empty set for L2 spaces.

File: src/utils.jl (hunk 1 above)
Comment (termi-official) [reply src/utils.jl:465]: Resolved with latest commit and generator fix in https://github.com/Ferrite-FEM/Ferrite.jl/pull/557

**Comment by termi-official** [comment]
Apart from test coverage for `interpolate_gradient_field` this one should be fine.

**Comment by koehlerson** [comment]
> Edit: 3D is future work when clip planes are implemented.

How does the docs example with 3D plasticity work then? :D

**Comment by termi-official** [comment]
Derp.

**Comment by koehlerson** [comment]
take the incompressible elasticity, there we can visualize both, stresses and strains, or alternatively the heat flux (so negative gradient with some constant) to show off how to get from gradient to flux

**Review by koehlerson — COMMENTED** [review]
few suggestions

**Comment by termi-official** [comment]
Okay, I think now only the next Ferrite release is missing (because of the L2 fix you pushed @koehlerson ).

**Comment by koehlerson** [comment]
> Okay, I think now only the next Ferrite release is missing (because of the L2 fix you pushed @koehlerson ).

builds locally with `linear` ansatz for the displacements :+1:

**Comment by koehlerson** [comment]
we can also think about having `Ferrite#master` as docs dependency

**Comment by termi-official** [comment]
I would vote for sticking with the safe route, manually upgrading at each release to capture regressions.

**Comment by termi-official** [comment]
Gradient field converges for the heat problem towards the analytical gradient in 2d

```julia
julia> maximum(u_grad) - π/2
0.0072785284134289086
```

and 3d

``julia
julia> maximum(u_grad) - π/2
0.01160556577741545
```

**Comment by koehlerson** [comment]
I think we could add a `Union` in the field argument such that either one gradient or multiple ones are computed. We should also provide the option to add a deformationfield to the `dh_grad`, the latter will be annoying for setting up the total DoF vector for `dh_grad`

**Comment by termi-official** [comment]
Before we start introducing more and more hotfixes, I think we should focus on reorganizing the code.

**Comment by termi-official** [comment]
I will skip the arrows plot in favor of #55 and #54 . Should be good to go (minus getting Ferrite master into test CI).

File: docs/src/atopics.md [hunk 3]
```diff
@@ -1,5 +1,57 @@
...
+Therefore, we provide the ability to interpolate the gradient into a piecewise discontinuous field via `FerriteViz.interpolate_gradient_field`.
+This function may be moved to Ferrite in the future.
+
+In this quick example we show how to visualize strains and stresses side-by-side
+```@example 1
+using Ferrite
+import FerriteViz
+using FerriteViz: ε
+import WGLMakie #activating the backend, switch to GLMakie or CairoMakie (for 2D) locally
+WGLMakie.set_theme!(resolution=(1500, 1600)) # hide
```
Comment (koehlerson) [inline docs/src/atopics.md:22]:
```suggestion

```

I think this should be taken out. On my laptop this looks nice, but on my workstation with the large monitor it looks super odd, can you open the docs and make a screenshot for me in the best case on both - laptop and workstation monitor?

File: CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ - Docs and helper for gradient field visualization based on interpolation [#51][github-51].
+   Currently only useful in 2D, because we have no clip plane feature to introspect the interior 
+   of 3D problems.
+ - Manufactured heat problem to test correctness of gradient field computation and as a 
+   helper to generate scalar-valued solutions with different ansatz [#51][github-51].
+
+### Modified
+ - Incompressible elasticity solver now takes the Ansatz functions and the actual material 
+   parameters instead of the poisson number the Ansatz functions [#51][github-51].
+
```
Comment (koehlerson) [inline CHANGELOG.md:20]:
```suggestion

### Fixed
- An unknown bug has been fixed, which computes the colorbar `(min,max)` wrong. Now the `max` is set to be `1.01` of `min` guaranteeing that the value is larger than `min` if close to zero.
```

File: src/utils.jl [hunk 5]
```diff
@@ -424,3 +424,76 @@ function dof_to_node(dh::Ferrite.AbstractDofHandler, u::Array{T,1}; field::Int=1
...
+        # of the gradient field
+        Ferrite.reinit!(cv, cell)
+
+        # Now we simply loop over all basis functions of the gradient field and evaluate the gradient
+        for i ∈ 1:Ferrite.getnbasefunctions(ip_gradient)
+            uᵉgradi = Ferrite.function_gradient(cv, i, uᵉ)
+            for ds in 1:dim
+                for df in 1:Ferrite.getfielddim(dh,field_name)
+                    # Note that we layout uᵉ_gradient as an order 3 tensor [num_base_funs, spatial_dim, field_dim]
+                    uᵉ_gradient[(i-1)*(dim*field_dim)+(ds-1)*field_dim+df] = uᵉgradi[ds,df]
```
Comment (koehlerson) [inline src/utils.jl:489]: no one on this godless earth will ever reconstruct the first index arithmetic

**Review by koehlerson — APPROVED** [review]
Otherwise LGTM

File: src/utils.jl [hunk 6]
```diff
@@ -467,26 +469,26 @@ function interpolate_gradient_field(dh::Ferrite.DofHandler{dim}, u::AbstractVect
...
         # Now we simply loop over all basis functions of the gradient field and evaluate the gradient
-        for i ∈ 1:Ferrite.getnbasefunctions(ip_gradient)
+        for i ∈ 1:num_base_funs
             uᵉgradi = Ferrite.function_gradient(cv, i, uᵉ)
-            for ds in 1:dim
+            for ds in 1:spatial_dim
                 for df in 1:Ferrite.getfielddim(dh,field_name)
-                    # Note that we layout uᵉ_gradient as an order 3 tensor [num_base_funs, spatial_dim, field_dim]
-                    uᵉ_gradient[(i-1)*(dim*field_dim)+(ds-1)*field_dim+df] = uᵉgradi[ds,df]
+                    uᵉ_gradient_view[df, ds, i] = uᵉgradi[ds, df]
```
[the remark this answers is not in the document]
Comment (termi-official) [reply src/utils.jl:491]: I think this is buggy now.

[the remark this answers is not in the document]
Comment (termi-official) [reply src/utils.jl]: Fixed.

**Comment by termi-official** [comment]
Correctness of shear component computation has been roughly tested via some crappy simple shear

```
include("docs/src/ferrite-examples/incompressible-elasticity.jl")

function create_cook_grid(nx, ny)
    grid = generate_grid(Triangle, (nx, ny));
    # facesets for boundary conditions
    addfaceset!(grid, "traction", x -> x[1] ≈ 2.0);
    return grid;
end;

function create_bc(dh)
    dbc = ConstraintHandler(dh)
    add!(dbc, Dirichlet(:u, getfaceset(dh.grid, "left"), (x,t) -> zero(Tensors.Vec{2}), [1,2]))
    add!(dbc, Dirichlet(:u, getfaceset(dh.grid, "right"), (x,t) -> Tensors.Vec{2}((0.0, 0.1)), [1,2]))
    close!(dbc)
    t = 0.0
    update!(dbc, t)
    return dbc
end;

u_quadratic,dh_quadratic = solve(quadratic, linear, mp);

(dh_quadratic_grad, u_quadratic_grad) = FerriteViz.interpolate_gradient_field(dh_quadratic, u_quadratic, :u)
plotter_quadratic = FerriteViz.MakiePlotter(dh_quadratic_grad, u_quadratic_grad)
cmap = :jet

f = GLMakie.Figure()
axs = [GLMakie.Axis(f[1, 1], title="u_{x}"),GLMakie.Axis(f[1, 2], title="u_{y}"),
       GLMakie.Axis(f[3, 1], title="u_{x,y}"),GLMakie.Axis(f[3, 2], title="u_{y,x}")]
p1 = FerriteViz.solutionplot!(axs[1], dh_quadratic, u_quadratic, process=u->u[1], colormap=cmap)
p2 = FerriteViz.solutionplot!(axs[2], dh_quadratic, u_quadratic, process=u->u[2], colormap=cmap)
f[2,1] = GLMakie.Colorbar(f[1,1], p1, vertical=false)
f[2,2] = GLMakie.Colorbar(f[1,2], p2, vertical=false)

p4 = FerriteViz.solutionplot!(axs[3], plotter_quadratic, process=∇u->∇u[2], colormap=cmap, colorrange = (-1e-2, 1e-2))
p5 = FerriteViz.solutionplot!(axs[4], plotter_quadratic, process=∇u->∇u[3], colormap=cmap, colorrange = (-0e-2, 6e-2))
f[4,1] = GLMakie.Colorbar(f[3,1], p4, vertical=false)
f[4,2] = GLMakie.Colorbar(f[3,2], p5, vertical=false)
```

Screenshot from 2023-02-13 18-07-44

**Comment by termi-official** [comment]
CI failed because it turns out that Tensors.jl gradients are the outermost indices, but access to a Vec (case: scalar problem) with two indices does not work. :) Should be fine now.

File: .github/workflows/docs.yml [hunk 7]
```diff
@@ -16,7 +16,7 @@ jobs:
         with:
           version: '1.6'
       - name: Install dependencies
-        run: julia --project=docs/ -e 'using Pkg; Pkg.develop(PackageSpec(path=pwd())); Pkg.instantiate()'
+        run: julia --project=docs/ -e 'using Pkg; Pkg.add(name="Ferrite",rev="master"); Pkg.develop(PackageSpec(path=pwd())); Pkg.instantiate()'
```
Comment (termi-official) [inline .github/workflows/docs.yml:19]: Should we revert this? @koehlerson

Comment (koehlerson) [reply .github/workflows/docs.yml:19]: yes

**Pull request merged**